File: cpp/search_in_rotated_sorted_array_ii.cc
/*
 * Filename: search_in_rotated_sorted_array_ii.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sun 02 Aug 2015 09:09:43 PM EDT
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		bool search(vector<int>& nums, int target) {
			if (nums.size() == 1) {
				return nums[0] == target ? true : false;
			}

			int len = nums.size();
			int offset = 0;
			int stt = 0;
			int end = nums.size() - 1;
			int mid = -1;

			for (int i = 1; i < len; i++) {
				if (nums[i] < nums[i-1]) {
					offset = i;
				}
			}

			while (stt <= end) {
				mid = (stt + end) / 2;
				int idx = (mid + offset) % len;
				
				if (nums[idx] == target) {
					return true;
				} else if (nums[idx] > target) {
					end = mid - 1;
				} else {
					stt = mid + 1;
				}
			}

			return false;
		}
};

File: java/SearchinRotatedSortedArrayII.java
/*
 * Filename: SearchinRotatedSortedArray.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Sun 22 Mar 2015 01:24:12 PM EDT
 */

public class Solution {
	public boolean search(int[] nums, int target) {
		int n = nums.length;

		if (n == 0) {
			return false;
		}

		if (n == 1) {
			if (nums[0] == target) {
				return true;
			} else {
				return false;
			}
		}

		int i = 1;
		while (i < n && nums[i-1] <= nums[i]) {
			i++;
		}

		int head = 0;
		int tail = n - 1;
		int mid;
		int val;
		
		while (head <= tail) {
			mid = (head + tail) / 2;
			val = nums[(mid + i) % n];
			if (val == target) {
				return true;
			} else if (val < target) {
				head = mid + 1;
			} else {
				tail = mid - 1;
			}
		}

		return false;
	}
}
